src/sql/src/names.rs

```diff
@@ -1131,6 +1171,244 @@ where
+        Either::Left(object_name)
+    }
+
+    fn fold_data_type(&mut self, data_type: A::DataType) -> <Either<A, B> as AstInfo>::DataType {
+        Either::Left(data_type)
+    }
+
+    fn fold_schema_name(&mut self, name: A::SchemaName) -> <Either<A, B> as AstInfo>::SchemaName {
+        Either::Left(name)
+    }
+
+    fn fold_database_name(
+        &mut self,
+        database_name: A::DatabaseName,
+    ) -> <Either<A, B> as AstInfo>::DatabaseName {
+        Either::Left(database_name)
+    }
+
+    fn fold_cluster_name(
+        &mut self,
+        cluster_name: A::ClusterName,
+    ) -> <Either<A, B> as AstInfo>::ClusterName {
+        Either::Left(cluster_name)
+    }
+
+    fn fold_with_option_value(
+        &mut self,
+        node: mz_sql_parser::ast::WithOptionValue<A>,
+    ) -> mz_sql_parser::ast::WithOptionValue<Either<A, B>> {
+        use mz_sql_parser::ast::WithOptionValue::*;
+        match node {
+            Value(v) => Value(Fold::<A, Either<A, B>>::fold_value(self, v)),
+            Ident(i) => Ident(Fold::<A, Either<A, B>>::fold_ident(self, i)),
+            DataType(dt) => DataType(Fold::<A, Either<A, B>>::fold_data_type(self, dt)),
+            Secret(secret) => Secret(Fold::<A, Either<A, B>>::fold_object_name(self, secret)),
+            Object(obj) => Object(Fold::<A, Either<A, B>>::fold_object_name(self, obj)),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EitherUnwrapper<F> {
+    inner: F,
+}
+
+impl<F> EitherUnwrapper<F> {
+    fn new(inner: F) -> Self {
+        Self { inner }
+    }
+}
+
+impl<A: AstInfo, B: AstInfo, F: Fold<B, A>> Fold<Either<A, B>, A> for EitherUnwrapper<F> {
+    fn fold_nested_statement(
+        &mut self,
+        stmt: <Either<A, B> as AstInfo>::NestedStatement,
+    ) -> A::NestedStatement {
+        match stmt {
+            Either::Left(stmt) => stmt,
+            Either::Right(stmt) => self.inner.fold_nested_statement(stmt),
+        }
+    }
+
+    fn fold_query(&mut self, q: Query<Either<A, B>>) -> Query<A> {
+        Query {
+            ctes: q.ctes.into_iter().map(|c| self.fold_cte(c)).collect(),
+            body: self.fold_set_expr(q.body),
+            limit: q.limit.map(|l| self.fold_limit(l)),
+            offset: q.offset.map(|l| self.fold_expr(l)),
+            order_by: q
+                .order_by
+                .into_iter()
+                .map(|c| self.fold_order_by_expr(c))
+                .collect(),
+        }
+    }
+
+    fn fold_cte_id(&mut self, id: <Either<A, B> as AstInfo>::CteId) -> A::CteId {
+        match id {
+            Either::Left(id) => id,
+            Either::Right(id) => self.inner.fold_cte_id(id),
+        }
+    }
+
+    fn fold_object_name(
+        &mut self,
+        object_name: <Either<A, B> as AstInfo>::ObjectName,
+    ) -> A::ObjectName {
+        match object_name {
+            Either::Left(object_name) => object_name,
+            Either::Right(object_name) => self.inner.fold_object_name(object_name),
```

Comment:
I think I don't fully understand the code yet, why does the right side always get folded but the left side never does?

Reply:
I'll add more comments but basically this folder turns an `Either<A, B>` into just `A` by running a `B -> A` folder over all `Right(B)` elements and reusing all `Left(A)` elements.

In the specific instantiation of `Either<Aug, Raw>` this means that if something is already resolved we keep it as-is and if something is `Raw` we resolve it